src/axom/quest/tests/quest_sampling_shaper.cpp

```diff
@@ -0,0 +1,868 @@
+    ::testing::UnitTest::GetInstance()->current_test_info()->name();
+
+  const std::string shape_template = R"(
+dimensions: 2
+units: cm
+
+shapes:
+- name: background
+  material: {0}
+  geometry:
+    format: none
+# Left replaces background void
+- name: left_side
+  material: {1}
+  geometry:
+    format: none
+# Odd cells replace background void, but not left
+- name: odd_cells
+  material: {2}
+  geometry:
+    format: none
+  does_not_replace: [{1}]
+)";
+
+  ScopedTemporaryFile shape_file(
+    axom::fmt::format("{}.yaml", testname),
+    axom::fmt::format(shape_template, "void", "left", "odds"));
+
+  if(very_verbose_output)
+  {
+    SLIC_INFO("Shape file: \n" << shape_file.getFileContents());
+  }
+
+  // Create an initial background material set to 1 everywhere
+  std::map<std::string, mfem::GridFunction*> initialGridFunctions;
+  {
+    auto* vf = this->registerVolFracGridFunction("init_vf_bg");
+    this->initializeVolFracGridFunction(
+      vf,
+      [](int, const Point2D&, int) -> double { return 1.; });
+    initialGridFunctions["void"] = vf;
+
+    // Note: element attributes were set earlier based on quadrant of cell's centroid (1, 2, 3 and 4)
+    vf = this->registerVolFracGridFunction("init_vf_left");
+    this->initializeVolFracGridFunction(
+      vf,
+      [](int, const Point2D&, int attr) -> double {
+        return (attr == 3 || attr == 4) ? 1. : 0;
+      });
+    initialGridFunctions["left"] = vf;
+
+    vf = this->registerVolFracGridFunction("init_vf_odds");
+    this->initializeVolFracGridFunction(
+      vf,
+      [](int idx, const Point2D&, int) -> double {
+        return idx % 2 == 1 ? 1. : 0;
+      });
+    initialGridFunctions["odds"] = vf;
+  }
+
+  this->validateShapeFile(shape_file.getFileName());
+  this->runShaping(shape_file.getFileName(), initialGridFunctions);
+
+  // check that the result has a volume fraction field associated with the circle material
+  const auto range = this->meshBoundingBox().range();
+
+  // Left covers half the mesh and is not replaced
+  const double expected_left_area = range[0] * range[1] / 2.;
+  // Odds should cover half of the right side of the mesh
+  const double expected_odds_area = range[0] * range[1] / 4.;
+  // The rest should be void
+  const double expected_void_area =
+    range[0] * range[1] - expected_left_area - expected_odds_area;
+
+  this->checkExpectedVolumeFractions("left", expected_left_area);
+  this->checkExpectedVolumeFractions("odds", expected_odds_area);
+  this->checkExpectedVolumeFractions("void", expected_void_area);
+
+  if(very_verbose_output)
+  {
+    this->getDC().Save(testname, axom::sidre::Group::getDefaultIOProtocol());
+  }
+}
+
+TEST_F(SamplingShaperTest, disk_with_multiple_preshaped_materials)
+{
+  using Point2D = typename SamplingShaperTest::Point2D;
+
+  const std::string& testname =
+    ::testing::UnitTest::GetInstance()->current_test_info()->name();
+
+  const std::string shape_template = R"(
```

Comment:
This is a complicated example. I'll post some screenshots from VisIt to show what it does. See also below for a textual description.